### PR TITLE
Fix evicting error and improve the related logs

### PR DESCRIPTION
**Pull request by shuo-wu**
https://github.com/longhorn/longhorn/issues/4294

### CODE REVIEW

File: controller/engine_controller.go [hunk 1]
```diff
@@ -1676,3 +1684,30 @@ func (ec *EngineController) isResponsibleFor(e *longhorn.Engine, defaultEngineIm
 
 	return isPreferredOwner || continueToBeOwner || requiresNewOwner, nil
 }
+
+func EngineOpStatusFilter(e *longhorn.Engine) {
+	tcpReplicaAddrMap := map[string]bool{}
```
Comment (c3y1huang) [inline controller/engine_controller.go:1689]: Suggest to use `map[string]struct{}` instead.

Comment (shuo-wu) [reply controller/engine_controller.go:1689]: Can you explain more about this? Will Golang apply any optimizations when using `struct{}`?

**Review by innobead — COMMENTED** [review]
LGTM, just some comments to resolve.

File: controller/engine_controller.go (hunk 1 above)
Comment (c3y1huang) [reply controller/engine_controller.go:1689]: I was looking where the boolean is used, and seems we are only using it for key lookups. AFAIK Using an empty struct doesn’t require memory allocated but a bool does.

File: controller/engine_controller.go [hunk 2]
```diff
@@ -863,11 +865,12 @@ func (m *EngineMonitor) refresh(engine *longhorn.Engine) error {
 
 	// Make sure the engine object is updated before engineapi calls.
 	if !reflect.DeepEqual(existingEngine.Status, engine.Status) {
-		engine, err = m.ds.UpdateEngineStatus(engine)
-		existingEngine = engine.DeepCopy()
-		if err != nil {
+		EngineOpStatusFilter(engine)
+
+		if engine, err = m.ds.UpdateEngineStatus(engine); err != nil {
```
[the remark this answers is not in the document]
Comment (shuo-wu) [reply controller/engine_controller.go:870]: No. That's why we use this function in the monitor thread only.

**Review by innobead — APPROVED** [review]

**Pull request merged**